### DISCUSSION TITLE
nlpsol() takes a long time

### DISCUSSION THREAD

**Discussion by waynekyrie** [casadi/casadi#3777] [Q&A] [answered]
Hi, I'm solving an optimization problem with a lot of constraints. Right now the initialization time for nlpsol() (~20s) is much slower than the solving time (~3s). Can someone provide any insights on how to improve the initialization time?

I looked into this post [https://github.com/casadi/casadi/issues/1958](url) and already have `{"ipopt.hessian_approximation":"limited-memory"}` to turn off exact Hessian. The post also mentioned to reformulate the problem to improve Hessian sparsity, but I'm not sure how to do this. Is there any examples? 

Below is the output from nlpsol()
```
Number of nonzeros in equality constraint Jacobian...:     1938
Number of nonzeros in inequality constraint Jacobian.:  1796214
Number of nonzeros in Lagrangian Hessian.............:        0

Total number of variables............................:      637
                     variables with only lower bounds:        0
                variables with lower and upper bounds:      427
                     variables with only upper bounds:        0
Total number of equality constraints.................:      487
Total number of inequality constraints...............:    15342
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:      210
        inequality constraints with only upper bounds:    15132
```

**Comment by tmmsartor** [comment]
Hello, it maybe be easier to give some suggestions if you provide more information about how your problem is formulated.
If you are solving an optimal control type of problem a general suggestion is to interleave state variables and constraints in order to get a more band-diagonal type of sparsity. I think the [video](https://youtu.be/X-nS4i8jJyM?si=jREgZii_Rzj_Ob9u) linked in this [demo](https://github.com/jgillis/fatrop_demo) does a great job at explaining this.
It is also likely that to decrease construction time you may need some other tricks.

**Reply by waynekyrie** [reply, marked as answer]
Thank you so much. I'm now just using code gen to store the problem and reuse it, which is acceptable in my case.

Just curious, is the FATROP in the link you sent available now in Python? If so, I would be really interested to try it. Thanks!

**Reply by tmmsartor** [reply]
Yes FATROP plugin in already shipping with Python interface, but not yet in the official release which is version 3.6.5.
You have to use the pre-release version 3.6.6 available on github https://github.com/casadi/casadi/releases/tag/nightly-release-3.6.6 . 
You have to download there right `.whl` file (it is a python wheel), i.e. if you are using linux with python 3.12 on intel cpu it would be  `casadi-3.6.6-cp312-none-manylinux2014_x86_64.whl `
then you can install it on your Python environment just by typing `pip install <file>.whl`

**Reply by tmmsartor** [reply]
Nevermind casadi 3.6.6 has just been [released](https://github.com/casadi/casadi/releases/tag/3.6.6), so now you can just type `pip install --upgrade casadi`

**Reply by waynekyrie** [reply]
Thank you so much!
